Add "Clean Slate" LangDef Boilerplate

tajmone opened the pull request:
New "`cleanslate.lang`" boilerplate: this langDef defines all possible syntax elements with never-matching RegExs, so that no elements ever match. Also overrides Highlight default fallback definitions. This allows to work with a clean slate approach, defining one syntax element at the time.

"`template.lang`" v1.0.2 (minor tweaks).

tajmone committed:
Add "Clean Slate" LangDef Boilerplate

New "cleanslate.lang" boilerplate: this langDef defines all possible syntax elements with never-matching RegExs, so that no elements ever match. Also overrides Highlight default fallback definitions. This allows to work with a clean slate approach, defining one syntax element at the time.

"template.lang" v1.0.2 (minor tweaks).

tajmone commented:
Hi André. I thought that this new "Clean Slate" boilerplate could be helpful. Full details in the source comments; but basically it proposes an approach where one starts with all syntax elements defined as never-matching RegExs, which ensures a "vacuum-like" environment in which one can focus on individual elements without other elements interfering.

I've used this boilerplate to solve conflicts while working on odd syntaxes (non-programming langs) where it was difficult to isolate some elements overriding a given definition. By switching between the never-matching definitions and my own definitions (via commenting/uncommenting lines) I was able to isolate what definitions where taking over the expected behavior.

Hopefully, others might find this useful (especially for odd syntaxes, like BNF-based syntaxes, markup syntaxes, etc.).

In due time I'll add also a nice html README file. I was also planning to add a document with reference examples — eg: comments and string delimiters definitions from various langs (C, Lua, etc.) — which might be handier than having to peek at all the existing langDef sources.

tajmone commented:
I've just realized that in the templates I've forgot to cover the `PreProcessor` element!

_(see Issue #45, where I explain the problem)_

As soon as I work out it's behavior I'll update both `template.lang` and `cleanslate.lang` accordingly.

andre-simon commented:
"Invalidating" `Digits` and `Identifiers` might be confusing as simple keyword lists will not work. The OnStateChange hook should be mentioned to point to a possible fix for parsing problems.
Another possibility would be to mention the never matching regex in the template.lang, since both files are very similar.

tajmone commented:
> "Invalidating" Digits and Identifiers might be confusing as simple keyword lists will not work. 

Right now I'm working at a new langDef which is an extended EBNF, in this case I didn't need `Digits`. Also, I realized that by disabling `Identifiers` it prevented some problems that I couldn't pinpoint. But it's true, I'm not using list-based keywords in this case, just RegEx keywords.

I've never understood the real purpose of `Identifiers` — this is not an element which gets tagged in the final output, so I guess it's for the parser's internal use. But what is it needed for? and how does changing it's definiton affect the langDef?

> simple keyword lists will not work.

Does this mean that `Identifiers` is the basic pattern for matching listed keywords? ie: that if a keyword is added that violates its pattern it will never be matched?

Going back to the EBNF definition I'm working on right now, all identifiers in this case are either non-terminal symbols or labels, which are handled by different keywords (RegEx based). I've noticed that stting `Identifiers` to never-match RegEx works fine, and actually removed some issues which were popping out for other elements (but then I might be confused on this by the multiple changes over time).

> The OnStateChange hook should be mentioned to point to a possible fix for parsing problems.

I'll add that, or at least add a short comment advising to check `template.lang` file which has a longer explanation (so maintainance is kept in a single file without redundancy).

> Another possibility would be to mention the never matching regex in the template.lang, since both files are very similar.

I did add it with this PR, in the headings:

```
-----------------------------------------------------------------------------
** HOW TO DISABLE SYNTAX ELEMENTS **
If you wish to suppress a syntax element, you can assign a never matching
regular expression to its definition (thanks to André Simon for the tip):

    [=[ \A(?!x)x ]=]
------------------------------------------------------------------------------
```

tajmone commented:
I've again come across the problem I mentioned with using the predefined value for `Identifiers`:

Unless I set `Identifiers` to a never matching RegEx, the parser fails to capture the underscore `_` operator!

It seems that HL's default RegEx interferes with my operators defintion:

```lua
Identifiers=[=[ [a-zA-Z_]\w* ]=] -- Highlight's default Identifiers definition

Operators=[=[ ::?=|;|:|,|\.|_|\\|\||\^|\+|-|>|<|\(|\)|\[|]\{|} ]=]
```

... the current syntax I'm working on (EBNF variant) uses underscore as "Epsilon" (empty production). Example:

``` ebnf
EPSILON ::= ball | _ ;
```

... it looks like HL interprets the standalone underscore as an Identifier, an somehow this prevents it from seeing it as an operator.

Of course, if I define `Identifiers` without underscore, it works again:

``` lua
Identifiers=[=[ [a-zA-Z]\w* ]=]
```

... but since I don't need `Digits` or `Identifiers` for this syntax, I'm better off just assigning them a never-matching RegEx (all keywords are RegEx defined, so they don't suffer from this).

In my view, this could be an argument in favour of Clean-Slate boilerplate giving `Identifiers` a never-matching RegEx as default (but with added warnings, and a reminder in keyword list), or at least as a commented out option. A total clean-slate approach would mean starting from scratch, with no default, at the cost of having to define `Identifiers` as one goes along. But maybe this could be confusing to some. This second template was more intended for experienced users, while the previous one for beginners.

I'm also thinking of adding a string variable (`NEVER_MATCH_RE`) defined once, at the beginning of the file, instead of repeating the full RegEx every time:

```lua
NEVER_MATCH_RE = [=[ \A(?!x)x ]=] -- A Never-Matching RegEx!

Digits = NEVER_MATCH_RE -- <= user var instead of RegEx!
```

... this would be cleaner, and also as a reminder to the user that some elements still lack proper RegEx definitions. (I know that in Lua all-capitals vars identifiers are usually for internals only, but in this context I'd want it to stand out since at the end of the work all reference to the variable should be gone).

tajmone committed:
Add PreProcessor to `template.lang`

Add PreProcessor presets, description and useful comments.

tajmone commented:
Based on your help from Issue #45, I've added to `template.lang` a `PreProcessor` example preset, the description from the documentation, and some help notes. After testing I've noticed that a further difference between singleline comments and `PreProcessor` is that in preprocessor lines the parser still looks for comments and strings — which makes sense. It seems to ignore the rest of the elements (eg `Digits`, keywords, etc.), but I haven't worked out what the parser considers an acceptable element within a preprocessor string, and what not.

tajmone committed:
Update "cleanslate.lang"

Identifiers with never-match RegEx is now uncommented by default, and a note explains cases when using a never-match RE might be useful.

Added:

- PreProcessor
- Reminder of OnStateChange() hooks

tajmone commented:
Also updated `cleanslate.lang` as you suggested.

But you might be right: having added the never-matching RegEx to `template.lang` might dispense of the need of this second boilerplate. I've proposed it because I find it useful to have some boilerplate to start working right out of the box on a new langDef (without having to copy and paste from all over the place); and of course I hate to have to reinvent the wheel every time.

For some jobs, I find this last boilerplate more handy thant the other template, but of course it depends on the task at hand, and of how one likes to work.

I leave it up to you to decide if it's worth adding or not. I thought sharing it was a good idea, and a sort of feedback on how Highlight users approach working on syntaxes.

Maybe once we're ready to merge I'll squash all these commits into one.

andre-simon commented:
Identifiers is the pattern for elements of keyword lists. Only tokens known as identifiers are searched in these lists. I assume that the underscore is not matched as operator if it is recognized as valid identifier first (precedence rules apply).

tajmone commented:
> Identifiers is the pattern for elements of keyword lists. Only tokens known as identifiers are searched in these lists. I assume that the underscore is not matched as operator if it is recognized as valid identifier first (precedence rules apply).

I had realized this. But what I don't understand is why the `_` was not matched as an operator: after all, there was no keywords list nor kewyords RegEx pattern that would match a standalone `_`. I thought it was just an issue of elements priorities, but it seems that a token matching the Identifiers patter is consumed by the parser even if doen't match any element in the keywords lists — I would have expected that if the token didn't match any keywords it would be passed on for checks against lower priority elements (ie: operators).

Instead, it looks like the Identifiers pattern behaves like a fishing net that always captures and consumes tokens.

It would be really useful to document the parser's flow, it's elements priorities and conditional branches. A diagram of its behavior could help solve syntax conflicts and implement fine-tuned hook functions. I'll try to look at the source code and jot down some notes on this.

andre-simon commented:
Identifiers can also be variable names which are not highlighted, like variables in C. So it it is not mandatory to be recognized as keyword, still they could be handled in plug-ins if needed. You got a rare syntax which allows '_' to be part of a keyword and a stand alone operator :)

tajmone commented:
Thanks! When I stumbled upon the underscore conflict in my syntax for PolyGen, it came to my mind that some languages do allow standalone `_`  to be an identifier (Python, if I'm not wrong, allows its use for internals and overrides; I have some vague memory of coming across this in a book).

Can you confirm me that identifiers are consumed by the parser (ie: won't be subjected to further checks), even if they don't match any keywords in the lists? Because this is the impression I got.

andre-simon commented:
Yes, a string which matches the Identifier regex will be treated as KEYWORD or STANDARD. This is handled here: https://github.com/andre-simon/highlight/blob/4a3788d7862c737796dc62f04f37ce730770a2c8/src/core/codegenerator.cpp#L567

andre-simon commented:
I have added a comment in the codegenerator header which gives a little insight about the parser: 

https://github.com/andre-simon/highlight/blob/0c975686dc40c452dd899700c88650214d8c356f/src/include/codegenerator.h#L110

Maybe still too terse but this is the general concept. It explains how the input lines are split into tokens, and why the syntax recognition might behave unexpectedly.

tajmone commented:
Thanks André, this really helps a lot!

When I'll put my hands again on the templates I'll both link to it and make us of it .

tajmone commented:
Added Wiki page:

https://github.com/tajmone/highlight/wiki/Parser-States
